akka-remote/src/main/scala/akka/remote/artery/Handshake.scala

```diff
@@ -197,21 +200,29 @@ private[akka] class InboundHandshake(inboundContext: InboundContext, inControlSt
         })
 
       private def onHandshakeReq(from: UniqueAddress): Unit = {
-        inboundContext.completeHandshake(from)
-        inboundContext.sendControl(from.address, HandshakeRsp(inboundContext.localAddress))
-        pull(in)
+        after(inboundContext.completeHandshake(from)) {
+          inboundContext.sendControl(from.address, HandshakeRsp(inboundContext.localAddress))
+          pull(in)
+        }
+      }
+
+      private def after(first: Future[Done])(thenInside: ⇒ Unit): Unit = {
```

Comment:
in the normal case (all but the first) the future will be completed, because handshake was already completed
note that we send those HandshakeReq periodically